test/langtools/tools/javac/patterns/DeconstructionDesugaring.java

```diff
@@ -0,0 +1,124 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8291769
+ * @summary Verify more complex switches work properly
+ * @compile --enable-preview -source ${jdk.version} DeconstructionDesugaring.java
+ * @run main/othervm --enable-preview DeconstructionDesugaring
+ */
+
+import java.util.function.ToIntFunction;
+public class DeconstructionDesugaring {
+
+    public static void main(String... args) throws Throwable {
+        new DeconstructionDesugaring().test();
+    }
+
+    private void test() {
+        test(this::runCheckStatement);
+        test(this::runCheckExpression);
+        assertEquals(runCheckExpressionWithUnconditional(new R5(new R4(new Sub3()))), 3);
+        assertEquals(runCheckExpressionWithUnconditional(new R5(new R4(null))), 3);
+        assertEquals(runCheckExpressionWithUnconditional1(new R5(new R4(null))), 2);
+        assertEquals(runCheckExpressionWithUnconditional1(new R5(null)), 3);
+    }
+
+    private void test(ToIntFunction<Object> task) {
+        assertEquals(1, task.applyAsInt(new R1(new R2(""))));
+        assertEquals(2, task.applyAsInt(new R1(new R2(1))));
+        assertEquals(3, task.applyAsInt(new R1(new R2(1.0))));
+        assertEquals(-1, task.applyAsInt(new R1(new R2(null))));
+        assertEquals(4, task.applyAsInt(new R1(new R2(new StringBuilder()))));
+        assertEquals(5, task.applyAsInt(new R1(new R3(""))));
+        assertEquals(6, task.applyAsInt(new R1(new R3(1))));
+        assertEquals(7, task.applyAsInt(new R1(new R3(1.0))));
+        assertEquals(8, task.applyAsInt(new R1(new R3(new StringBuilder()))));
+        assertEquals(-1, task.applyAsInt(new R1(1.0f)));
+        assertEquals(-1, task.applyAsInt("foo"));
+    }
+
+    private int runCheckStatement(Object o) {
+        switch (o) {
+            case (((R1((((R2((((String s))))))))))) -> { return 1; }
+            case R1(R2(Integer i)) -> { return 2; }
+            case R1(R2(Double d)) -> { return 3; }
+            case R1(R2(CharSequence cs)) -> { return 4; }
+            case R1(R3(String s)) -> { return 5; }
+            case R1(R3(Integer i)) -> { return 6; }
+            case R1(R3(Double f)) -> { return 7; }
+            case R1(R3(CharSequence cs)) -> { return 8; }
+            default -> { return -1; }
+        }
+    }
+
+    private int runCheckExpression(Object o) {
+        return switch (o) {
+            case (((R1((((R2((((String s))))))))))) -> 1;
```

Comment:
general notes on the tests, I think that we need to add records with varargs components as in:
```
record R(int... indices) {}
```
and records with annotations on their components that are also preprocessed with annotation processors.